python/rmm/rmm/_lib/memory_resource.pyx

```diff
@@ -723,6 +749,43 @@ cdef class LimitingResourceAdaptor(UpstreamResourceAdaptor):
             self.c_obj.get())
         )[0].get_allocation_limit()
 
+cdef class SamHeadroomResourceAdaptor(UpstreamResourceAdaptor):
+    def __cinit__(
+        self,
+        DeviceMemoryResource upstream_mr,
+        size_t headroom
+    ):
+        if not isinstance(upstream_mr, SystemMemoryResource):
```

Comment:
Does `SystemMemoryResource` maintain any per-instance state, or does it simply call malloc/free? If the latter, then there's no need to pass an upstream resource to this constructor at all, the constructor could create one.

Reply:
Note also that this requirement makes it impossible to use an *adapted* system memory resource as upstream (e.g. logging, aligned, statistics, tracking etc.). Not a huge issue, just noting it.

Reply:
Done.